src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp

```diff
@@ -15,751 +15,879 @@
+                if (_events.IsInPhase(PHASE_THREE))
+                    DoCast(me, SPELL_ELECTRICAL_CHARGE, true);
+                break;
+            case ACTION_BERSERK_TRIGGERED:
+                Talk(SAY_STEELBREAKER_BERSERK);
+                break;
+            default:
+                break;
+        }
+    }
 
-            void UpdateAI(uint32 diff) override
-            {
-                if (!UpdateVictim())
-                    return;
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim() || !CheckInRoom())
+            return;
 
-                events.Update(diff);
+        _events.Update(diff);
 
-                if (me->HasUnitState(UNIT_STATE_CASTING))
-                    return;
+        if (me->HasUnitState(UNIT_STATE_CASTING) || me->HasSpellFocus())
+            return;
 
-                while (uint32 eventId = events.ExecuteEvent())
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_FUSION_PUNCH:
+                    DoCastVictim(SPELL_FUSION_PUNCH);
+                    _events.Repeat(13s, 22s);
+                    break;
+                case EVENT_STATIC_DISRUPTION:
                 {
-                    switch (eventId)
-                    {
-                        case EVENT_BERSERK:
-                            Talk(SAY_STEELBREAKER_BERSERK);
-                            DoCast(SPELL_BERSERK);
-                            events.CancelEvent(EVENT_BERSERK);
-                            break;
-                        case EVENT_FUSION_PUNCH:
-                            if (me->IsWithinMeleeRange(me->GetVictim()))
-                                DoCastVictim(SPELL_FUSION_PUNCH);
-                            events.ScheduleEvent(EVENT_FUSION_PUNCH, 13s, 22s);
-                            break;
-                        case EVENT_STATIC_DISRUPTION:
-                            if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0))
-                                DoCast(target, SPELL_STATIC_DISRUPTION);
-                            events.ScheduleEvent(EVENT_STATIC_DISRUPTION, 20s, 40s);
-                            break;
-                        case EVENT_OVERWHELMING_POWER:
-                            Talk(SAY_STEELBREAKER_POWER);
-                            DoCastVictim(SPELL_OVERWHELMING_POWER);
-                            events.ScheduleEvent(EVENT_OVERWHELMING_POWER, RAID_MODE(60s, 35s));
-                            break;
-                    }
-
-                    if (me->HasUnitState(UNIT_STATE_CASTING))
-                        return;
+                    // Steelbreaker prefers ranged targets above melee targets but falls back to melee targets if no ranged target is available
+                    Unit* target = SelectTarget(SelectTargetMethod::Random, 0, -10.f);
+                    if (!target)
+                        target = SelectTarget(SelectTargetMethod::Random);
```

Comment:
any chance of this still returns null after that?
maybe this update tick happening in a wipe?

Reply:
No, because in that case, the threatlist would be empty so the boss should not be engaged at all.